File: Kelly/queue.linkedList/queue.linkedList.ts
class Node<T> {
  value: T;
  next: Node<T> | null;

  constructor(value: T) {
    this.value = value;
    this.next = null;
  }
}

class Queue<T> {
  private head: Node<T> | null;
  private tail: Node<T> | null;
  private size: number;

  constructor() {
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  enqueue(value: T): void {
    const newNode = new Node<T>(value);

    if (this.isEmpty()) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail!.next = newNode;
      this.tail = newNode;
    }

    this.size++;
  }

  dequeue(): T | null {
    if (this.isEmpty()) {
      return null;
    }

    const removedNode = this.head!;
    this.head = this.head!.next;
    removedNode.next = null;

    if (this.head === null) {
      this.tail = null;
    }

    this.size--;

    return removedNode.value;
  }

  peek(): T | null {
    if (this.isEmpty()) {
      return null;
    }

    return this.head!.value;
  }

  isEmpty(): boolean {
    return this.size === 0;
  }

  getSize(): number {
    return this.size;
  }

  printQueue(): void {
    let current = this.head;

    while (current !== null) {
      console.log(current.value);
      current = current.next;
    }
  }
}

File: Kelly/queue.linkedList/queue.linkedList.cs
public class Queue<T>
{
    private class Node
    {
        public T Value { get; }
        public Node Next { get; set; }

        public Node(T value)
        {
            Value = value;
        }
    }

    private Node _head;
    private Node _tail;
    private int _count;

    public int Count => _count;
    public bool IsEmpty => _count == 0;

    public void Enqueue(T item)
    {
        Node newNode = new Node(item);

        if (IsEmpty)
        {
            _head = newNode;
            _tail = newNode;
        }
        else
        {
            _tail.Next = newNode;
            _tail = newNode;
        }

        _count++;
    }

    public T Dequeue()
    {
        if (IsEmpty)
        {
            throw new InvalidOperationException("Queue is empty.");
        }

        T value = _head.Value;
        _head = _head.Next;
        _count--;

        if (IsEmpty)
        {
            _tail = null;
        }

        return value;
    }

    public T Peek()
    {
        if (IsEmpty)
        {
            throw new InvalidOperationException("Queue is empty.");
        }

        return _head.Value;
    }

    public void Clear()
    {
        _head = null;
        _tail = null;
        _count = 0;
    }
}
